Source/Details/ASTraitCollection.m

```diff
@@ -32,36 +85,58 @@ extern void ASTraitCollectionPropagateDown(id<ASLayoutElement> element, ASPrimit
     .userInterfaceIdiom = UIUserInterfaceIdiomUnspecified,
+    .layoutDirection = UITraitEnvironmentLayoutDirectionUnspecified,
+    .preferredContentSizeCategory = ASPrimitiveContentSizeCategoryMake(AS_AT_LEAST_IOS10 ? UIContentSizeCategoryUnspecified : AS_UIContentSizeCategoryUnspecified),
     .containerSize = CGSizeZero,
   };
 }
 
-ASPrimitiveTraitCollection ASPrimitiveTraitCollectionFromUITraitCollection(UITraitCollection *traitCollection)
-{
+ASPrimitiveTraitCollection ASPrimitiveTraitCollectionFromUITraitCollection(UITraitCollection *traitCollection) {
   ASPrimitiveTraitCollection environmentTraitCollection = ASPrimitiveTraitCollectionMakeDefault();
-  environmentTraitCollection.displayScale = traitCollection.displayScale;
   environmentTraitCollection.horizontalSizeClass = traitCollection.horizontalSizeClass;
   environmentTraitCollection.verticalSizeClass = traitCollection.verticalSizeClass;
+  environmentTraitCollection.displayScale = traitCollection.displayScale;
   environmentTraitCollection.userInterfaceIdiom = traitCollection.userInterfaceIdiom;
   if (AS_AT_LEAST_IOS9) {
     environmentTraitCollection.forceTouchCapability = traitCollection.forceTouchCapability;
   }
+  if (AS_AT_LEAST_IOS10) {
+    environmentTraitCollection.displayGamut = traitCollection.displayGamut;
+    environmentTraitCollection.layoutDirection = traitCollection.layoutDirection;
+
+    // preferredContentSizeCategory is also available on older iOS versions, but only via UIApplication class.
+    // It should be noted that [UIApplication sharedApplication] is unavailable because Texture is built with only extension-safe API.
+    environmentTraitCollection.preferredContentSizeCategory = ASPrimitiveContentSizeCategoryMake(traitCollection.preferredContentSizeCategory);
+
+    #if TARGET_OS_TV
+      environmentTraitCollection.userInterfaceStyle = traitCollection.userInterfaceStyle;
+    #endif
+  }
   return environmentTraitCollection;
 }
 
-BOOL ASPrimitiveTraitCollectionIsEqualToASPrimitiveTraitCollection(ASPrimitiveTraitCollection lhs, ASPrimitiveTraitCollection rhs)
-{
+BOOL ASPrimitiveTraitCollectionIsEqualToASPrimitiveTraitCollection(ASPrimitiveTraitCollection lhs, ASPrimitiveTraitCollection rhs) {
   return
     lhs.verticalSizeClass == rhs.verticalSizeClass &&
     lhs.horizontalSizeClass == rhs.horizontalSizeClass &&
     lhs.displayScale == rhs.displayScale &&
+    lhs.displayGamut == rhs.displayGamut &&
     lhs.userInterfaceIdiom == rhs.userInterfaceIdiom &&
     lhs.forceTouchCapability == rhs.forceTouchCapability &&
+    lhs.layoutDirection == rhs.layoutDirection &&
+    #if TARGET_OS_TV
+      lhs.userInterfaceStyle == rhs.userInterfaceStyle &&
+    #endif
+
+    // preferredContentSizeCategory always points to one of UIContentSizeCategory constants.
+    // Assuming their values do not duplicate, we can simply compare pointers, avoiding string comparison.
+    lhs.preferredContentSizeCategory == rhs.preferredContentSizeCategory &&
```

Comment:
Thank you for the detailed comment. I actually expect `-[NSString isEqualToString:]` to compare memory addresses early on. Without concrete numbers, I'd consider this optimization unnecessary, also because calling `-isEqualToString:` gives us better maintainability.

Interesting find: https://gist.github.com/0xced/2275014#file-nsstring-m-L76